Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs

```diff
@@ -203,6 +333,324 @@ public static DualShockHIDOutputReport Create()
+                   currentState->buttons2 == nextState->buttons2;
+        }
+
+        unsafe bool IEventMerger.MergeForward(InputEventPtr currentEventPtr, InputEventPtr nextEventPtr)
+        {
+            if (currentEventPtr.type != StateEvent.Type || nextEventPtr.type != StateEvent.Type)
+                return false;
+        
+            var currentEvent = StateEvent.FromUnchecked(currentEventPtr);
+            var nextEvent = StateEvent.FromUnchecked(nextEventPtr);
+        
+            if (currentEvent->stateFormat != DualSenseHIDGenericInputReport.Format || nextEvent->stateFormat != DualSenseHIDGenericInputReport.Format)
+                return false;
+
+            if (currentEvent->stateSizeInBytes != nextEvent->stateSizeInBytes)
+                return false;
+
+            var currentGenericReport = (DualSenseHIDGenericInputReport*)currentEvent->state;
+            var nextGenericReport = (DualSenseHIDGenericInputReport*)nextEvent->state;
+
+            if (currentGenericReport->reportId != nextGenericReport->reportId)
+                return false;
+
+            if (currentGenericReport->reportId == 0x01)
+            {
+                if (currentEvent->stateSizeInBytes == 10 || currentEvent->stateSizeInBytes == 78)
+                {
+                    var currentState = (DualSenseHIDMinimalInputReport*)currentEvent->state;
+                    var nextState = (DualSenseHIDMinimalInputReport*)nextEvent->state;
+                    return MergeForward(currentState, nextState);
+                }
+                else
+                {
+                    var currentState = (DualSenseHIDUSBInputReport*)currentEvent->state;
+                    var nextState = (DualSenseHIDUSBInputReport*)nextEvent->state;
+                    return MergeForward(currentState, nextState);
+                }
+            }
+            else if (currentGenericReport->reportId == 0x31)
+            {
+                var currentState = (DualSenseHIDBluetoothInputReport*)currentEvent->state;
+                var nextState = (DualSenseHIDBluetoothInputReport*)nextEvent->state;
+                return MergeForward(currentState, nextState);
+            }
+            else
+                return false;
+        }
+        
+        unsafe bool IEventPreProcessor.PreProcessEvent(InputEventPtr eventPtr)
+        {
+            if (eventPtr.type != StateEvent.Type)
+                return eventPtr.type != DeltaStateEvent.Type; // only skip delta state events
+
+            var stateEvent = StateEvent.FromUnchecked(eventPtr);
+            var size = stateEvent->stateSizeInBytes;
+            if (stateEvent->stateFormat != DualSenseHIDGenericInputReport.Format || size < sizeof(DualSenseHIDInputReport))
+                return false; // skip unrecognized state events otherwise they will corrupt control states
+
+            var genericReport = (DualSenseHIDGenericInputReport*)stateEvent->state;
+            if (genericReport->reportId == 0x01)
```

Comment:
Could we put the magic numbers behind some well named constants here? 0x01, 0x31, and the two state event byte sizes 10 and 78

Reply:
fixed